ocaml/typing/rec_check.ml

```diff
@@ -1192,13 +1192,24 @@ and case
        ----------------------------------------
        G - p; m[mp] |- (p (when g)? -> e) : m
     *)
-    let judg = join [
-        option expression c_guard << Dereference;
-        expression c_rhs;
-      ] in
-    (fun m ->
-       let env = judg m in
-       (remove_pat c_lhs env), Mode.compose m (pattern c_lhs env))
+      let judg = match c_guard with
+        | None -> expression c_rhs
+        | Some (Predicate p) -> join [ expression p << Dereference; expression c_rhs ]
+        | Some (Pattern (e, _, pat)) -> pattern_guard e pat c_rhs in
+      (fun m ->
+        let env = judg m in
+        (remove_pat c_lhs env), Mode.compose m (pattern c_lhs env))
+
+and pattern_guard
```

Comment:
Can we refactor this to more explicitly share the code that's run for `Texp_match`? If we can't do that, we'll need to comment `pattern_guard` in the same way that other functions in this file are comment (with its corresponding judgment and derivation rules thereof).

Reply:
Done!